Docs for testing ZenML pipelines (#2250)
zenml-io/zenml: docs/book/user-guide/advanced-guide/best-practices/unit_testing.md

# Write unit testing
In ZenML, unit testing plays a crucial role in ensuring the robustness and reliability of machine learning pipelines. Sometimes, it can be beneficial to extract common functionalities into separate functions to prevent code duplication and streamline the pipeline development process. This approach not only enhances code readability but also simplifies the testing phase by isolating specific functionalities for testing.

## Why unit testing is required?
Unit testing is essential for ensuring the reliability and correctness of individual components within pipelines. It serves as a quality assurance mechanism, validating that each step, like data loaders or transformations does their job as intended in isolation. Unit tests provide an additional layer of confidence by verifying specific functionalities, catching bugs early in the development cycle, and enabling easier troubleshooting. By confirming that each unit operates correctly independent of the entire pipeline, it promotes robustness, enhances maintainability, and fosters a more efficient development process within ZenML.

> ## How is unit testing is tested?

Reviewer: ```suggestion
## How to unit test ZenML steps?
```

Unit testing in ZenML can be carried out by writing test cases using frameworks like Pytest. These test cases focus on individual units, such as steps within pipelines, to validate their functionality in isolation. Test functions are created to check specific behaviors or inputs and ensure expected outputs. Mocking and patching are often used to simulate dependencies or external interactions, allowing thorough testing without affecting the actual environment. Continuous integration and deployment (CI/CD) pipelines further automate the execution of these tests, running them regularly to guarantee ongoing code stability and reliability in ZenML's development cycle.

> For more practical view, this is our ZenML step code:

Reviewer: ```suggestion
For a more practical view, let's say this is out ZenML step:

```python
@step(enable_cache=False)
def step_2(input_one: str, input_two: str) -> None:
    """Combines the two strings passed in."""
    combined_str = f"{input_one} {input_two}"
    return combined_str
```

> This is a ZenML step, and ensuring its proper functionality is crucial. In case it fails, you want to detect this issue early to prevent potential disasters. To achieve this, writing unit tests for each function is necessary. While you have the flexibility to structure functions based on your needs, it's advisable to follow a one-test-case-per-function approach. This practice aligns with the principles of unit testing, focusing on individual functionalities rather than integration. When running these unit test functions, the terminal displays the count of passed test cases.

Reviewer: ```suggestion
As this is a step within a pipeline, ensuring its proper functionality is crucial. In case it fails, we want to detect early to prevent potential disasters. It's advisable to follow a one-test-case-per-function approach. This practice aligns with the principles of unit testing, focusing on individual functionalities.
```

> So, here is the unit testing for your ZenML step code written above.

Reviewer: ```suggestion
So, here is the corresponding unit test of the above step:
```

```Python
def test_step_2():
    input_one = "Hello"
    input_two = "World"

>     import test_step #file name

Reviewer: We can also import the step at the top of the file, this might give the wrong impression of importing things within tests

                                             steps=[self.split_mock,
                                                    self.preprocess_mock,
                                                    self.trainer_mock,
                                                    self.evaluator_mock])
        
        # Set isolation for each step in the pipeline
        for step in training_pipeline.steps:
            self.assertTrue(step.isolation, msg=f"{step} is not isolated.")
            
    def test_step_isolation(self):
        # Create individual steps
        split = RandomSplit()
        preprocess = StandardScaler()
        trainer = TFTrainer()
        evaluator = TFEvaluator()
        
        # Set isolation for each step individually
        split.isolation = True
        preprocess.isolation = True
        trainer.isolation = True
        evaluator.isolation = True
        
        self.assertTrue(split.isolation, msg="RandomSplit step is not isolated.")
        self.assertTrue(preprocess.isolation, msg="StandardScaler step is not isolated.")
        self.assertTrue(trainer.isolation, msg="TFTrainer step is not isolated.")

>         self.assertTrue(evaluator.isolation, msg="TFEvaluator step is not isolated.")

Reviewer: This is wrong syntax for ZenML pipelines, datasource and steps are not an input, and overall this seems like old GPT generated ;-)

Author: Oops, Got caught  .sorry for that I need to grow / change my habit . I'll fix that out :)
